pkg/git/v2/transport.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2020 The Flux CD contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v2
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/sha1"
+	"fmt"
+	"golang.org/x/crypto/ssh"
+	"strings"
+
+	"github.com/fluxcd/source-controller/pkg/git/common"
+	git2go "github.com/libgit2/git2go/v31"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func AuthSecretStrategyForURL(url string) common.AuthSecretStrategy {
+	switch {
+	case strings.HasPrefix(url, "http"):
+		return &BasicAuth{}
+	case strings.HasPrefix(url, "ssh"):
+		return &PublicKeyAuth{}
+	}
+	return nil
+}
+
+type BasicAuth struct{}
+
+func (s *BasicAuth) Method(secret corev1.Secret) (*common.Auth, error) {
+	var username string
+	if d, ok := secret.Data["username"]; ok {
+		username = string(d)
+	}
+	var password string
+	if d, ok := secret.Data["password"]; ok {
+		password = string(d)
+	}
+	if username == "" || password == "" {
+		return nil, fmt.Errorf("invalid '%s' secret data: required fields 'username' and 'password'", secret.Name)
+	}
+
+	credCallback := func(url string, username_from_url string, allowed_types git2go.CredType) (*git2go.Cred, error) {
+		cred, err := git2go.NewCredUserpassPlaintext(username, password)
+		if err != nil {
+			return nil, err
+		}
+		return cred, nil
+	}
+
+	return &common.Auth{CredCallback: credCallback, CertCallback: nil}, nil
```

Comment:
Does `CertCallback: nil` means that we accept invalid certs?

Reply:
Good question, I read the docs last week about this but is good that you bring it up so that if this becomes a question in the future we can link here. 

This is what the libgit2 docs state.
>If cert verification fails, this will be called to let the user make the final decision of whether to allow the connection to proceed. Returns 0 to allow the connection or a negative value to indicate an error.

https://libgit2.org/libgit2/#HEAD/type/git_remote_callbacks

My interpretation of this is that the cert callback is just an additional verification step to enable custom checks that extend that which is in the OS CA.

The same interpretation is stated in a pygit2 issue.
https://github.com/libgit2/pygit2/issues/627#issuecomment-214679069

Reply:
We should add a test with a self-signed cert to make sure libgit2 rejects it, I think there is a test for Helm repos certs cc @hiddeco

Reply:
There are test certificates in `controllers/testdata/certs`.